### ISSUE TITLE
Devices transition plugged -> dead -> plugged is still not completely fixed

### ISSUE THREAD

**Issue by fbuihuu** [systemd/systemd#12953]
This happened on Tumbleweed which ships systemd v242.

For the complete debug logs, please consult https://bugzilla.suse.com/show_bug.cgi?id=1137373.

Except from the logs:

> [   25.202778] fyrfos systemd[1]: Switching root.
> [   36.627399] fyrfos systemd[1]: Starting udev Coldplug all Devices...
> [   36.714237] fyrfos systemd[1]: Started udev Coldplug all Devices.
> [   48.182070] fyrfos systemd[1]: Reloading.
> [   48.433421] fyrfos systemd[1]: Invoking unit coldplug() handlers…
> [   48.457773] fyrfos systemd[1]: dev-sdb5.device: Changed dead -> plugged
> [   48.466512] fyrfos systemd[1]: Invoking unit catchup() handlers…
> [   48.471053] fyrfos systemd[1]: dev-sdb5.device: Changed plugged -> dead
> [   48.741803] fyrfos systemd[1]: dev-sdb5.device: Changed dead -> plugged
> [   49.005134] fyrfos systemd[1]: local.mount: local.mount lost dependency BindsTo=dev-sdb5.device
> 

This was already reported in https://github.com/systemd/systemd/issues/11997 but the fix was incorrect, see https://github.com/systemd/systemd/pull/12013#issuecomment-479038739

**Comment by fbuihuu** [comment]
I think the regression is severe enough to deserve the v243 milestone.

**Comment by yuwata** [comment]
I've reopend the previous issue page...

**Comment by fbuihuu** [comment]
Damn I missed that and strangely enough I haven't received any notification about your comment in the other bug although I'm subscribed to it.

Maybe we should continue the discussion here ?

**Comment by yuwata** [comment]
OK, old one is closed again.

**Comment by fbuihuu** [comment]
@yuwata I started working on a possible fix for this, please have a look at: https://github.com/fbuihuu/systemd/commits/fix-inconsistent-device-state-transition

**Comment by yuwata** [comment]
@fbuihuu Do you have a reproducer of this issue?

**Comment by fbuihuu** [comment]
@yuwata no sorry. I tried to run "systemctl daemon-reload" right before `systemd-udevd.service` is started but it led to a broken system and I haven't found some time to investigate why.

**Comment by poettering** [comment]
since this isn't clear year, let's move to v244

**Comment by fbuihuu** [comment]
See https://github.com/systemd/systemd/pull/13002 for more details.

**Comment by fbuihuu** [comment]
@yuwata, it seems that commit c6e892bc0eebe1d42c282bd2d8bae149fbeba85f also introduced a regression.

IIRC the commit removed the "plugged -> dead -> plugged" transition after switching root and devices in "plugged" state in initramfs stay in "plugged" state after switching root.

But this prevents device dependencies which are only defined in the root filesystem (i.e. not included in initramfs) to not be pulled in and started.

**Comment by fbuihuu** [comment]
It would nice if all of these (new) issues could be fixed for good.

**Comment by fbuihuu** [comment]
@keszybz, IMHO we shouldn't postpone this (again).

**Comment by fbuihuu** [comment]
Another user affected by this issue: https://bugzilla.opensuse.org/show_bug.cgi?id=1155170

**Comment by fbuihuu** [comment]
It would be great if such bug could be fixed now.

**Comment by keszybz** [comment]
We should return to this at some point...

**Comment by fbuihuu** [comment]
we also should stop postponing this at every new release ;)

**Comment by fkru** [comment]
It seems that I am also affected by this bug on openSUSE Tumbleweed, i.e., swap partition is activated/deactivated during boot by systemd, leaving me at the end with an unmounted swap partition: https://bugzilla.suse.com/show_bug.cgi?id=1165946

**Comment by Daniele71** [comment]
Any news ?

**Comment by dsvi** [comment]
Is there a patch available for that?

**Comment by fbuihuu** [comment]
@poettering what is the point to set a milestone label if it keeps getting postponed to the next release ?

Using a milestone label in the first place means that it's an important (design) bug/regression which we should really fix..

**Comment by keszybz** [comment]
> @poettering what is the point to set a milestone label if it keeps getting postponed to the next release ?

We move items to a later milestone if they turn out to be to complex to fix. So something being assigned to a milestone means "yes, we'll not forget about this and we'll try to implement it before release, but it may get bumped if nobody works on it or there is no progress for other reasons".

**Comment by fbuihuu** [comment]
That seems to defeats the purpose of milestones... I mean deferring the milestone for one release is fine, but it's been deferred 4 times already. At that point it's simply better to drop it.

**Comment by yuwata** [comment]
But, how should we fix the issue? At least I have no idea to completely fix it.

**Comment by poettering** [comment]
So, what about this approach to fix this all in PID1:

1. We'll start serializing .device state across reload/restart/switch-root, and then deserialize our state from that after
2. We never consult the udev db anymore
3. We only act on udev uevents for tracking device state
4. We make the `systemd-udev-trigger.service` ExecStart= line synchronous: by adding the `--settle` switch to the `udevadm trigger` command line the service will remain in in "activating" state until all devices are triggered and their uevents propagated through the system. 
5. We make `systemd-udev-trigger.service` special in PID 1: when we are about to start it we start our udev uevents monitor, and when we are finished stopping it we close our udev uevent monitor. Thus the lifetime of this service also describes the time when PID 1 listens to uevents. Also, once start-up of the service is complete PID 1 will iterate through its .device units and deactivate those which it didn't get a suitable uevent for in the service's start-up time.

This means if the udev db is incomplete/flushed it doesn't matter anymore to us: we only operate on "live" data, on real uevents as they happen, and we are "frozen" at the time the s-u-t.s service isn't running and thus do not risk undoing thing the initrd has set up for us.

**Comment by poettering** [comment]
`udevadm trigger --settle` probably needs a time-out btw, to make this robust

**Comment by poettering** [comment]
Given 47cd17ead4bd773691708f3e81a7d047b5692e51 I think this is not as pressing anymore, hence bumping again.

**Comment by arvidjaar** [comment]
> We only act on udev uevents for tracking device state
> the lifetime of this service [`systemd-udev-trigger.service`]also describes the time when PID 1 listens to uevents

This sounds like systemd will only process uevents when `systemd-udev-trigger.service` is running or may be I misunderstand something?

Otherwise this sounds logical. udevd is not needed to use device. Device is there when kernel says device is there. udevd may collect device properties and it may ask systemd to delay followup processing because device may not be yet considered ready but it should be opt in.

**Comment by poettering** [comment]
> This sounds like systemd will only process uevents when `systemd-udev-trigger.service` is running or may be I misunderstand something?

Yes, exactly.

**Comment by fbuihuu** [comment]
> So, what about this approach to fix this all in PID1:

IIRC the problem is due to the fact that udev DB is cleared when the system switches to rootfs.

Can't we try to address that instead by either letting udev run while switching to rootfs (that's what we recommend for storage daemons after all) or saving safely any uevents that could occur until udev is started again ?

See https://github.com/systemd/systemd/pull/13002#issuecomment-578032090.

**Comment by poettering** [comment]
passing the udev db over would mean that host and initrd run the same version and the same comprehensive set of udev rules. That's generally not the case though: the udev ruleset on the host is typically a lot more comprehensive than the set included in the initrd, might even container local admin additions and might come from a different udev/distro version. 

Or to say this differently: the data passed from initrd to host should be very very carefully picked, minimal and stable. Blanket passing the whole udev database over is the wrong approach.

**Comment by fbuihuu** [comment]
Isn't one of the reason why systemd-udev-trigger.service is started in the host ? 

So that the udev DB is updated with the new set of udev rules.

**Comment by hexiaowen** [comment]
https://github.com/systemd/systemd/issues/21220
It's the same problem, no doubt.

**Comment by syyhao1994** [comment]
https://github.com/systemd/systemd/issues/21214
The problem mentioned in the issue above also is the same problem.

**Comment by syyhao1994** [comment]
I add some debug info. So the process of the early daemon-reload action could be observed more clearly.

The service i used is list as below, i made it before local-fs-pre.target for making this service activated as soon as possible after switch root.
```
[root@localhost ~]# systemctl cat firstrun.service
# /usr/lib/systemd/system/firstrun.service
[Unit]
Description=First run service
Before=local-fs-pre.target

[Service]
Type=oneshot
ExecStart=/usr/bin/systemctl daemon-reload
KillMode=process

[Install]
WantedBy=multi-user.target
```
The process could simplified as below
```
Nov 04 10:12:04 localhost.localdomain systemd[1]: Reloading.
...
Nov 04 10:12:04 localhost.localdomain systemd[1]: Invoking unit coldplug() handlers…
...
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into unit_coldplug
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into unit_coldplug
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into device_coldplug, the d->deserialized_state is 2,the d->state is 0, the d->found is 0, the d->deserialized_found is 1
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into device_set_state
Nov 04 10:12:04 localhost.localdomain systemd[1]: dev-sdb.device: Changed dead -> plugged
...
Nov 04 10:12:04 localhost.localdomain systemd[1]: Invoking unit catchup() handlers…
...
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy the unit is dev-sdb.device, The enumerated_found is 0, the found is 1
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into device_update_found_one,the device is (null),the d->found is 1,the d->enumerated_found is 0, the found is 0, the mask is 7
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into device_found_changed, the previous is 1,the now is 0
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy device_found_changed to dead
Nov 04 10:12:04 localhost.localdomain systemd[1]: syy into device_set_state
Nov 04 10:12:04 localhost.localdomain systemd[1]: dev-sdb.device: Changed plugged -> dead
...
then unmounted the device.
```
The debug info patch could see in https://github.com/syyhao1994/files/blob/master/add_debug_log.patch.

**Comment by arvidjaar** [comment]
@syyhao1994 

> ```
> Nov 04 15:59:21 systemd[1]: Reloading.
> Nov 04 15:59:21 systemd[1]: syy into device_coldplug, the d->deserialized_state is 2,the d->state is 0, the d->found is 0, the d->deserialized_found is 1
> Nov 04 15:59:21 systemd[1]: dev-cpsVG-fsp.device: Changed dead -> plugged
> ```

So before reload device was found by udev and was plugged.

> ```
> Nov 04 15:59:21 systemd[1]: syy the unit is dev-cpsVG-fsp.device, The enumerated_found is 0, the found is 1
> Nov 04 15:59:21 systemd[1]: dev-cpsVG-fsp.device: Changed plugged -> dead
> Nov 04 15:59:21 systemd[1]: home-fsp.mount: Trying to enqueue job home-fsp.mount/stop/replace
> ```

And that is a problem. Device was not found during initial enumeration. Even device is not yet in udev database, it must have been found by mount enumeration. My best guess is that `/proc/self/mountinfo` lists some other (canonical) path for this device which is different from alias used to mount it. Please, provide full log as well as output of `grep /home/fsp /proc/self/mountinfo`.

**Comment by arvidjaar** [comment]
I believe I at least understand the code flow. BTW it is trivial to reproduce it:

```
systemctl stop systemd-udevd.service
rm -rf /run/udev
systemctl start systemd-udevd.service
systemctl daemon-reload
```

So first `systemd` performs enumeration; it correctly creates unit for mounted device node with `d->enumerated_found == DEVICE_FOUND_MOUNT`.

Next it runs coldplug which succeeds to restore device's `d->found` to `DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV`.

Next `systemd` calls `device_catchup`. Because `d->enumerated_found == DEVICE_FOUND_MOUNT` is not equal to `d->found == DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV` we call `device_update_found_one(d, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MASK)` which calls `device_found_changed(d, DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV, DEVICE_FOUND_MOUNT)`.

Looking at this function - it does not expect "downgrade" of device presence. Previous state is not `DEVICE_NOT_FOUND`, `now` is not `DEVICE_NOT_FOUND`, but `DEVICE_FOUND_UDEV` flag is set in `previous` and  systemd assumes that this cannot happen and device is gone.

```
static void device_found_changed(Device *d, DeviceFound previous, DeviceFound now) {
        assert(d);

        /* Didn't exist before, but does now? if so, generate a new invocation ID for it */
        if (previous == DEVICE_NOT_FOUND && now != DEVICE_NOT_FOUND)
                (void) unit_acquire_invocation_id(UNIT(d));

        if (FLAGS_SET(now, DEVICE_FOUND_UDEV))
                /* When the device is known to udev we consider it plugged. */
                device_set_state(d, DEVICE_PLUGGED);
        else if (now != DEVICE_NOT_FOUND && !FLAGS_SET(previous, DEVICE_FOUND_UDEV))
                /* If the device has not been seen by udev yet, but is now referenced by the kernel, then we assume the
                 * kernel knows it now, and udev might soon too. */
                device_set_state(d, DEVICE_TENTATIVE);
        else
                /* If nobody sees the device, or if the device was previously seen by udev and now is only referenced
                 * from the kernel, then we consider the device is gone, the kernel just hasn't noticed it yet. */
                device_set_state(d, DEVICE_DEAD);
}
```

**Comment by syyhao1994** [comment]
> I believe I at least understand the code flow. BTW it is trivial to reproduce it:
> 
> ```
> systemctl stop systemd-udevd.service
> rm -rf /run/udev
> systemctl start systemd-udevd.service
> systemctl daemon-reload
> ```
> 
> So first `systemd` performs enumeration; it correctly creates unit for mounted device node with `d->enumerated_found == DEVICE_FOUND_MOUNT`.
> 
> Next it runs coldplug which succeeds to restore device's `d->found` to `DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV`.
> 
> Next `systemd` calls `device_catchup`. Because `d->enumerated_found == DEVICE_FOUND_MOUNT` is not equal to `d->found == DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV` we call `device_update_found_one(d, DEVICE_FOUND_MOUNT, DEVICE_FOUND_MASK)` which calls `device_found_changed(d, DEVICE_FOUND_MOUNT|DEVICE_FOUND_UDEV, DEVICE_FOUND_MOUNT)`.
> 
> Looking at this function - it does not expect "downgrade" of device presence. Previous state is not `DEVICE_NOT_FOUND`, `now` is not `DEVICE_NOT_FOUND`, but `DEVICE_FOUND_UDEV` flag is set in `previous` and systemd assumes that this cannot happen and device is gone.
> 
> ```
> static void device_found_changed(Device *d, DeviceFound previous, DeviceFound now) {
>         assert(d);
> 
>         /* Didn't exist before, but does now? if so, generate a new invocation ID for it */
>         if (previous == DEVICE_NOT_FOUND && now != DEVICE_NOT_FOUND)
>                 (void) unit_acquire_invocation_id(UNIT(d));
> 
>         if (FLAGS_SET(now, DEVICE_FOUND_UDEV))
>                 /* When the device is known to udev we consider it plugged. */
>                 device_set_state(d, DEVICE_PLUGGED);
>         else if (now != DEVICE_NOT_FOUND && !FLAGS_SET(previous, DEVICE_FOUND_UDEV))
>                 /* If the device has not been seen by udev yet, but is now referenced by the kernel, then we assume the
>                  * kernel knows it now, and udev might soon too. */
>                 device_set_state(d, DEVICE_TENTATIVE);
>         else
>                 /* If nobody sees the device, or if the device was previously seen by udev and now is only referenced
>                  * from the kernel, then we consider the device is gone, the kernel just hasn't noticed it yet. */
>                 device_set_state(d, DEVICE_DEAD);
> }
> ```

I understand what the **device_found_changed** do. But the log that i added indicated that the previous is DEVICE_FOUND_UDEV with out DEVICE_FOUND_MOUNT and now is DEVICE_NOT_FOUND. 
```
Nov 04 15:59:21 systemd[1]: syy into device_found_changed, the previous is 1,the now is 0
```
I am also confused about the value of **now** that i think it should be DEVICE_FOUND_MOUNT. But i think it is not the key point for this problem as it makes no difference whether the DEVICE_FOUND_MOUNT flag bit is set or not. This problem is caused by the **previous** num is with DEVICE_FOUND_UDEV but **now** without DEVICE_FOUND_UDEV.

**Comment by syyhao1994** [comment]
https://github.com/systemd/systemd/pull/21627
@keszybz @poettering @yuwata @fbuihuu 
Could please take a look at the PR mentioned above.
I think the problem introduced is much more serious and common than we want to fix. Could we please disable manager_catchup in daemon reload util we come up with a better method to do this.

**Comment by mwilck** [comment]
@arvidjaar's [analysis](https://github.com/systemd/systemd/issues/12953#issuecomment-962585519) makes a lot of sense to me. I think we shouldn't be clearing `DEVICE_FOUND_UDEV` from `device_catchup()`. If the manager is just reloaded, the socket listening for uevents will be kept open; outstanding uevents will be received after the reload. There's no need to derive updates for `d->state` or `d->found` from device enumeration; in particular no need to assume that a device that wasn't enumerated by udev has actually removed during the reload operation. 

`daemon-reexec` (and switching root) is different, but it uses different logic anyway (`m->honor_device_enumeration` is set to true after `manager_catchup()`)

Everyone please have a look at https://github.com/mwilck/systemd/commit/f01f8cbc14130e78e6e659e75bf6262fbb22a90f.

**Comment by mwilck** [comment]
One more thing: as pointed out by @arvidjaar in https://github.com/systemd/systemd/issues/12953#issuecomment-962585519, systemd currently doesn't deal well with device state "downgrade". It basically only supports full device removal ("plugged -> dead") which usually implies unmounting dependent file systems. The device activation logic could be changed, but it's complex and needs to cover a broad range of scenarios, thus changing it is likely to cause regressions.

The advantage of my proposal is that is doesn't need to modify the device activation logic. It just changes the *initial state* in one special situation (switching root) and relies on the already existing logic for handling "tentative" devices to do the right thing.
